src/Sentry/Internal/Http/CachingTransport.cs

```diff
@@ -0,0 +1,266 @@
+                    );
+                }
+            }
+
+            _worker = Task.Run(async () =>
+            {
+                while (!_workerCts.IsCancellationRequested)
+                {
+                    try
+                    {
+                        await _workerSignal.WaitAsync(_workerCts.Token).ConfigureAwait(false);
+                        await ProcessCacheAsync(_workerCts.Token).ConfigureAwait(false);
+                    }
+                    catch (OperationCanceledException)
+                    {
+                        // Ignore
+                    }
+                    catch (Exception ex)
+                    {
+                        _options.DiagnosticLogger?.LogError(
+                            "Exception in background worker of CachingTransport.",
+                            ex
+                        );
+                    }
+                }
+            });
+        }
+
+        private IEnumerable<string> GetCacheFilePaths()
+        {
+            if (!Directory.Exists(_cacheDirectoryPath))
+            {
+                return Enumerable.Empty<string>();
+            }
+
+            return Directory
+                .EnumerateFiles(_cacheDirectoryPath, $"*.{EnvelopeFileExt}")
+                .OrderBy(f => new FileInfo(f).CreationTimeUtc);
+        }
+
+        public int GetCacheLength() => GetCacheFilePaths().Count();
+
+        // Gets the next cache file and moves it to "processing"
+        private async ValueTask<string?> TryPrepareNextCacheFileAsync(
+            CancellationToken cancellationToken = default)
+        {
+            using var lockClaim = await _cacheDirectoryLock.ClaimAsync(cancellationToken).ConfigureAwait(false);
+
+            var filePath = GetCacheFilePaths().FirstOrDefault();
+            if (string.IsNullOrWhiteSpace(filePath))
+            {
+                return null;
+            }
+
+            var targetFilePath = Path.Combine(_processingDirectoryPath, Path.GetFileName(filePath));
+
+            Directory.CreateDirectory(_processingDirectoryPath);
+            File.Move(filePath, targetFilePath);
+
+            return targetFilePath;
+        }
+
+        private async ValueTask ProcessCacheAsync(CancellationToken cancellationToken = default)
+        {
+            _options.DiagnosticLogger?.LogDebug("Flushing cached envelopes.");
+
+            while (await TryPrepareNextCacheFileAsync(cancellationToken).ConfigureAwait(false) is { } envelopeFilePath)
+            {
+                _options.DiagnosticLogger?.LogDebug(
+                    "Reading cached envelope: {0}",
+                    envelopeFilePath
+                );
+
+                var shouldRetry = false;
+
+                using (var envelopeFile = File.OpenRead(envelopeFilePath))
+                using (var envelope =
+                    await Envelope.DeserializeAsync(envelopeFile, cancellationToken).ConfigureAwait(false))
```

Comment:
If the file got corrupted, isn't deserialization going to throw an Exception and the remaining Envelopes are not going to be loaded?

Reply:
Good point.